### ISSUE TITLE
Reduce number of orphan blocks resulting from merged mining

### ISSUE THREAD

**Issue by jansegre** [HathorNetwork/hathor-core#91]
Currently there are two sets of mining APIs:

- the first mining API, does not support merged mining:
  - `GET /v1a/mining`
  - `POST /v1a/mining`
- the current mining API, that resembles Bitcoin's, and is pretty complete:
  - `GET /v1a/get_block_template`
  - `POST /v1a/submit_block`

However, unlike Bitcoin's `getblocktemplate`, our `get_block_template` API does not support "longpolling" to wait for a new block height. Additionally `submit_block` could optimistically return a block template based on the submitted block (when valid).

These two features should greatly reduce the waiting period between a block being found and miners being notified that they should start working on a new job. Which in turn reduces the number of orphan (void) blocks.

I propose adding three JSONRPC ~2.0~ 1.0 (changed because it's easier to re-use JSONRPC 1.0 abstractions already used for Bitcoin, we can add support for both in the future) calls over WebSocket:

- The WebSocket endpoint is `/v1a/mining_ws`;
- The main RPC is a notification, `method=mining.notify`, that is sent by the fullnode (acting as a client, because that's how notifications are sent), where `params` is a list of block templates (described bellow);
- A `method=mining.submit` called by the miner (fullnode is now the server), it accepts two params (named or positional), `hexdata` and `optimistic` (in that order), `optimistic` being optional. `hexdata` is a proposed block serialized in hexadecimal (just like `/v1a/submit_block` accepts), ~and `optimistic` is a boolean that when true, makes the call~ return a block template ~(instead of true)~ on success that uses the submitted block as a parent (it's optimistic in the sense that it hopes this block will become the tip of the best chain and so facilitates starting mining on that sooner), if any error happens the result is ~`false`~ `null`;
- A `method=mining.refresh` called by the miner, the result is the same list of block templates that `mining.notification` would have as params. This is useful in case the network goes stale and the current timestamp becomes greater than all `timestamp_max` of the last block templates given, which can help triggering a difficulty decay as soon as possible;
- A block template structure is a "dict"/"object" with the following fields:
  - `versions`: possible values for a proposed blocks `version` field;
  - `reward`: integer amount of the reward value that should be used on the proposed block's output (6400 means 64.00 HTR);
  - `weight`: float number of the difficulty that should go as is into a proposed block;
  - `timestamp_min` and `timestamp_max`: the range that the `timestamp` field of the proposed block has to be to fit all other given parameters (like `weight` and `parents` which can vary depending on the `timestamp`)
  - `parents`: list of hashes (in hexadecimal) of the parents that must be included in the proposed block, has a most 3 elements, will always have at least on element (the block);
  - `parents_any`: if `parents` has less than 3 elements, this is a list of hashes that can be chosen from to fill the list and make it to 3
  - `height` and `score`: this is a metadata field (the height of the block and the total score of the chain the block is in), that can be useful for a miner to know, `height` can be used to set the `clean` parameter downstream, and `score` can be used if a miner is connected to multiple Hathor fullnodes and select the template with highest score;

Old version:

> Since we don't have a JSONRPC API, and eitherway using longpolling is a bit cumbersome for both the client and the server. I propose a WebSocket endpoint, that would do just that:
> 
> - ~`WS /v2/mining`~ `WS /v1a/mining_ws` (accepts the same parameters as `/v1a/get_block_template`):
>   - instead of yielding a single block template, yields a dict (json object) with:
>     - `cur_timestamp`: current fullnode timestamp, for reference
>     - `sync_timestamp`: from p2p layer, makes it so the MM server can connect to multiple fullnodes and choose de templates from the one with latest `sync_timestamp`
>     - `block_templates`: list of templates based on all best block tips, instead of randomly selecting one, MM servers are free to use whichever strategy for tip selection, additionally a block template will contain `min_timestamp` and `max_timestamp` instead of a `timestamp`, this is to signal which range of timestamp can be freely used for a new block with the given parameters (implementation-wise it is perfectly fine for this API to return templates with `min_timestamps` that are only usable in the future, this could be used to preemptively send blocks with decayed weight, but this doesn't have to be implemented right now, this is only a feature this format supports)
>   - accepts a refresh command that the server can use if the blockchain is stale and there were no tips changes and some of the `max_timestamp`s above have "expired"
>   - accepts submissions with the same format that `/v1a/submit_block` accepts, and optionally (`optimistic_best_chain` param) return a new block template based off the block just submitted (if valid)

**Comment by obiyankenobi** [comment]
>  tick every second (the server should change the timestamp to update the weight when needed)

Would our full nodes be responsible for this? Doesn't seem to me like it's their responsibility.

We can also connect the MM server through the current ws endpoint that sends blocks/txs in realtime.

**Comment by jansegre** [comment]
> > tick every second (the server should change the timestamp to update the weight when needed)
> 
> Would our full nodes be responsible for this? Doesn't seem to me like it's their responsibility.

It's only useful for one reason, but it's an important reason to me. A block's timestamp field isn't exactly freely changeable. For prolonged and sudden hashrate drops, the difficulty decay kicks in, and a change in timestamp would warrant a change in the transactions weight.

Resending the same block every second with a new timestamp (and potentially new weight) is the stupid simple way of solving this.

On second thought, there is a smarter way to do this, we could send a `min_timestamp` and `max_timestamp` (which would be the last valid timestamp before a decay) instead of a `timestamp`. The MM server is then free to choose a timestamp in that range, or not a timestamp higher than `max_timestamp` isn't a hard error just a waste of hashrate, but for stability of the network it is a very important feature that should work well, specially through the primary "hashrate channel". I'll update the issue with this change.

> We can also connect the MM server through the current ws endpoint that sends blocks/txs in realtime.

That is possible. My main reasoning against this is that the current ws channel is mostly concerned with wallet stuff. Mixing the two can make cross bugs more likely, but more importantly it can make it harder to gate or rate limit. And I don't really see a practical benefit to minimizing ws endpoints.

**Comment by jansegre** [comment]
We could bikeshed the proposed endpoint though. Some other alternatives are:

- `/v1a/mining/ws`
- `/v1a/ws/mining`

**Comment by obiyankenobi** [comment]
My concern about this is having the full node responsible for mining logic. We are migrating to an architecture where the full node will basically validate the txs and all other functions are removed from it (tx-mining-service, wallet-service). I think the same should happen with block mining, specially since we already have the MM coordinator as a separate code.

Your second suggestion seems more reasonable, though.

**Comment by jansegre** [comment]
> My concern about this is having the full node responsible for mining logic.

I completely agree. I could argue about what is mining logic, but I also think that a good API is ergonomic and maintainable, and in this case it could have a bit of mining logic.

In theory an efficient miner could be made on top of APIs that enable access to the blocks including signaling events for when new tx/blocks arrive. This would undoubtedly be free of any mining logic. But it would mean the miner has to re-implement block and tx tip selection logic and the DAA for calculating the next weight. I don't think we have to go to this extreme. Although we could in the future consider something very modular that would implement a mining API this way.

But I agree the originally proposed "ticking" was bad. I think this new API is a good compromise and also very ergonomic.

**Comment by pedroferreira1** [comment]
I think the first suggestion has been edited, what makes little bit hard to follow the discussion. But seeing the issue as it is now, looks good to me.

> But it would mean the miner has to re-implement block and tx tip selection logic and the DAA for calculating the next weight. I don't think we have to go to this extreme. 

I completely agree with this.

I just don't see any reason why we should move to `/v2/`, I would go with one of your suggestions to keep `/v1a/`.

So just to be clear, the moment a new block arrives, increasing the network height, a new message on the websocket would be sent, so all MM servers can update their mining data?

**Comment by jansegre** [comment]
> So just to be clear, the moment a new block arrives, increasing the network height, a new message on the websocket would be sent, so all MM servers can update their mining data?

Yes, but more than that. If a new block arrives, and the best chain tips change, then there will be an update too. The height doesn't have to change for an update to be sent. The miner will get all the best block tips and their heights so that it can implement whatever tip choosing logic it wants. For starters I will keep the current "random tip" logic, but this will now move from the fullnode to the MM server.

**Comment by obiyankenobi** [comment]
This new proposal seems good.

> additionally a block template will contain min_timestamp and max_timestamp instead of a timestamp

Is this `min_timestamp` and `max_timestamp` something particular to each block template? Or it should be the same for all?

**Comment by jansegre** [comment]
> Is this `min_timestamp` and `max_timestamp` something particular to each block template? Or it should be the same for all?

Each block template has a pair, it varies on the timestamp of the parent block.

**Comment by obiyankenobi** [comment]
> optimistic is a boolean that when true, makes the call return a block template (instead of true) on success that uses the submitted block as a parent (it's optimistic in the sense that it hopes this block will become the tip of the best chain and so facilitates starting mining on that sooner), if any error happens the result is false;

Is there a need for `optimistic`? Why is it not the default behavior? The client can choose to ignore the returned template instead of setting it to False.

> versions: possible values for a proposed blocks version field;

I don't understand this. What would the possible versions be?

> parents_any: if parents has less than 3 elements, this is a list of hashes that can be chosen from to fill the list and make it to 3

What's the benefit of having the miner choose parents? Is it something useful now or that we're adding for possible future use?

**Comment by jansegre** [comment]
> Is there a need for `optimistic`? Why is it not the default behavior? The client can choose to ignore the returned template instead of setting it to False.

It could actually be the default and only way. Probably the reasons I made it separate was to save some bandwidth and processing (which seems pretty much negligible now that I think about it).

> I don't understand this. What would the possible versions be?

Currently 1 (regular block) or 3 (merged mining block). This indicates this template can be used to make both types of blocks. We could deprecate this in the future, or evolve when the `version` field becomes more complex. At the very least it's useful for not messing up the generated block version.

> What's the benefit of having the miner choose parents? Is it something useful now or that we're adding for possible future use?

The main purpose is returning a predictable result. That is, not automatically picking random parents when available. The alternative would be choosing the parents randomly and dropping `parents_any`, which when refreshing could change and could invalidate caches when only parents change (but it's not that the parents changed just a different random was picked). But honestly the impact seems low, but I still feel like it's a better design to delegate the choice and have predictable (and arguably easier to test) API, at virtually no cost (the miner now has to do random choice, or if by any chance it prefers to chose one over the other, that is now possible).

**Comment by obiyankenobi** [comment]
> It could actually be the default and only way. Probably the reasons I made it separate was to save some bandwidth and processing (which seems pretty much negligible now that I think about it).

That was actually something I was thinking later. Creating the block template is pretty cheap?

> But honestly the impact seems low, but I still feel like it's a better design to delegate the choice and have predictable (and arguably easier to test) API, at virtually no cost

Got it, seems good to me.

**Comment by jansegre** [comment]
> > It could actually be the default and only way. Probably the reasons I made it separate was to save some bandwidth and processing (which seems pretty much negligible now that I think about it).
> 
> That was actually something I was thinking later. Creating the block template is pretty cheap?

I haven't benchmarked or profiled anything yet, but `make_block_template` should be reasonably light, which is called by `make_block_templates` which may be expensive when `storage.get_best_block_tips` isn't cached, but anyway `get_block_templates` caches the result anyway (which is invalidated when `storage.latest_timestamp` changes, which by the way is a good reason not to use random choices for the parents so the cache makes more sense), so calling `get_block_templates` a lot should be pretty cheap

**Comment by pedroferreira1** [comment]
`storage.get_best_block_tips` is updated every time a new transaction/block arrives, so it will always have a cache. The only moment we would need to calculate the tips is if you call the method with an specific timestamp (which is not the case, for what I've seen in the code). So yeah, it's supposed to be pretty cheap.

**Comment by jansegre** [comment]
The last bit we need to fix this is being discussed on #152
